testing/async/mixin.h

```diff
@@ -0,0 +1,645 @@
+ *
+ * template <typename input_value_type,
+ *           typename output_value_type = input_value_type>
+ * struct host_synchronous
+ * {
+ *   template <typename InputType,
+ *             typename OutputType,
+ *             typename PostfixArgTuple,
+ *             std::size_t... PostfixArgIndices>
+ *   static void invoke_reference(InputType const& input,
+ *                                OutputType& output,
+ *                                PostfixArgTuple&& postfix_tuple,
+ *                                std::index_sequence<PostfixArgIndices...>)
+ *   {
+ *     // Create host versions of the input/output:
+ *     thrust::host_vector<input_value_type> host_input(input.cbegin(),
+ *                                                      input.cend());
+ *     thrust::host_vector<output_value_type> host_output(host_input.size());
+ *
+ *     // Run host synchronous algorithm to generate reference.
+ *     // Be sure to call a backend that doesn't use the same underlying
+ *     // implementation.
+ *     thrust::exclusive_scan(host_input.cbegin(),
+ *                            host_input.cend(),
+ *                            host_output.begin(),
+ *                            std::get<PostfixArgIndices>(
+ *                              THRUST_FWD(postfix_tuple))...);
+ *
+ *     // Copy back to device.
+ *     output = host_output;
+ *   }
+ * };
+ *
+ */
+
+// Used to save time when testing unverifiable invocations (discard_iterators)
+struct noop
+{
+  template <typename... Ts>
+  static void invoke_reference(Ts&&...)
+  {}
+};
+
+} // namespace invoke_reference
+
+//------------------------------------------------------------------------------
+namespace invoke_async
+{
+
+/* Defined per algorithm. Example:
+ *
+ * struct basic
+ * {
+ *   template <typename PrefixArgTuple,
+ *             std::size_t... PrefixArgIndices,
+ *             typename InputType,
+ *             typename OutputType,
+ *             typename PostfixArgTuple,
+ *             std::size_t... PostfixArgIndices>
+ *   static auto invoke_async(PrefixArgTuple&& prefix_tuple,
+ *                            std::index_sequence<PrefixArgIndices...>,
+ *                            InputType const& input,
+ *                            OutputType& output,
+ *                            PostfixArgTuple&& postfix_tuple,
+ *                            std::index_sequence<PostfixArgIndices...>)
+ *   {
+ *     auto e = thrust::async::exclusive_scan(
+ *       std::get<PrefixArgIndices>(THRUST_FWD(prefix_tuple))...,
+ *       input.cbegin(),
+ *       input.cend(),
+ *       output.begin(),
+ *       std::get<PostfixArgIndices>(THRUST_FWD(postfix_tuple))...);
+ *     return e;
+ *   }
+ * };
+ */
+
+} // namespace invoke_async
+
+//------------------------------------------------------------------------------
+namespace compare_outputs
+{
+
+namespace detail
+{
+
+void basic_event_validation(thrust::device_event& e)
+{
+  TEST_EVENT_WAIT(e);
```

Comment:
Do we still need the macros if we're just abstracting this into a function? Can we nuke the macros?

Reply:
That sounds like a great candidate for a follow-up PR.